gnomad_constraint/utils/constraint.py

```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
+        possible_ht.annotate(mu_agg=possible_ht.mu_snp * possible_ht.possible_variants)
+    )
+    # Apply plateau and coverage models for all sites.
+    total_plateau_model = hl.literal(plateau_models.total)[possible_ht.cpg]
+    ann_expr = {
+        "predicted_proportion_observed": possible_ht.mu_agg * total_plateau_model[1]
+        + total_plateau_model[0],
+        "coverage_correction": hl.case()
+        .when(possible_ht.coverage == 0, 0)
+        .when(possible_ht.coverage >= cov_cutoff, 1)
+        .default(
+            coverage_model[1] * hl.log10(possible_ht.coverage) + coverage_model[0]
+        ),
+    }
+    # Apply plateau models for specified populations.
+    for pop in pops:
+        pop_model = hl.literal(plateau_models[pop])
+        slopes = hl.map(lambda f: f[possible_ht.cpg][1], pop_model)
+        intercepts = hl.map(lambda f: f[possible_ht.cpg][0], pop_model)
+        ann_expr[f"predicted_proportion_observed_{pop}"] = (
+            possible_ht.mu_agg * slopes + intercepts
+        )
+    possible_ht = possible_ht.annotate(**ann_expr)
+    # Compute expected variant counts for all sites.
+    ann_expr = {
+        "expected_variants": possible_ht.predicted_proportion_observed
+        * possible_ht.coverage_correction,
+        "mu": possible_ht.mu_agg * possible_ht.coverage_correction,
+    }
+    # Compute expected variant counts for specified populations.
+    for pop in pops:
+        ann_expr[f"expected_variants_{pop}"] = (
+            possible_ht[f"predicted_proportion_observed_{pop}"]
+            * possible_ht.coverage_correction
+        )
+    possible_ht = possible_ht.annotate(**ann_expr)
+
+    # Outer join the Tables with possible variant counts and observed variant counts.
+    ht = observed_ht.join(possible_ht, "outer")
+
+    grouping = list(grouping)
+    grouping.remove("coverage")
+
+    # Generate sum aggregators for 'observed_variants',
+    # 'predicted_proportion_observed', 'possible_variants', 'expected_variants', and
+    # 'mu' on the entire dataset.
+    agg_expr = {
+        "observed_variants": hl.agg.sum(ht.observed_variants),
+        "predicted_proportion_observed": hl.agg.sum(ht.predicted_proportion_observed),
+        "possible_variants": hl.agg.sum(ht.possible_variants),
+        "expected_variants": hl.agg.sum(ht.expected_variants),
+        "mu": hl.agg.sum(ht.mu),
+    }
+    # Generate sum aggregators for 'observed_variants',
+    # 'predicted_proportion_observed', 'possible_variants', 'expected_variants', and
+    # 'mu' for specified populations.
+    for pop in pops:
+        agg_expr[f"predicted_proportion_observed_{pop}"] = hl.agg.array_sum(
+            ht[f"predicted_proportion_observed_{pop}"]
+        )
+        agg_expr[f"expected_variants_{pop}"] = hl.agg.array_sum(
+            ht[f"expected_variants_{pop}"]
+        )
+        agg_expr[f"downsampling_counts_{pop}"] = hl.agg.array_sum(
+            ht[f"downsampling_counts_{pop}"]
+        )
```

Comment:
I think this can be simplified by creating just aggregations without the annotations in between. If they are needed for computational reasons then it's fine to keep annotations, but can still probably be made less redundant and easier to read.

Here is the solution with removing the annotation steps. Obviously things need to be cleaned up, but I think this is enough to show how it can be done.

```suggestion
    mu_expr = ht.mu_snp * ht.possible_variants

    # Determine coverage correction to use based on coverage value
    cov_corr_expr = (
        hl.case()
          .when(ht.coverage == 0, 0)
          .when(ht.coverage >= cov_cutoff, 1)
          .default(coverage_model[1] * hl.log10(ht.coverage) + coverage_model[0])
    )

    # Generate sum aggregators for 'mu' on the entire dataset.
    agg_expr = {"mu": hl.agg.sum(mu_expr * cov_corr_expr)}
    agg_expr.update(name_me(ht, plateau_models, mu_expr, cov_corr_expr))

    for pop in pops:
        agg_expr.update(name_me(ht, plateau_models, mu_expr, cov_corr_expr, pop))

    # Aggregate the sum aggregators grouped by `grouping`
    grouping = list(grouping)
    grouping.remove("coverage")
```

Add this function:
```
def name_me(ht, plateau_models, mu_expr, cov_corr_expr, pop: Optional[str] = None):
    if pop is None:
        pop = ""
        plateau_model = hl.literal(plateau_models.total)[ht.cpg]
        slope = plateau_model[1]
        intercept = plateau_model[0]
        agg_func = hl.agg.sum
        ann_to_sum = ["observed_variants", "possible_variants"]
    else:
        plateau_model = hl.literal(plateau_models[pop])
        slope = hl.map(lambda f: f[ht.cpg][1], plateau_model)
        intercept = hl.map(lambda f: f[ht.cpg][0], plateau_model)
        agg_func = hl.agg.array_sum
        pop = f"_{pop}"
        ann_to_sum = [f"downsampling_counts{pop}"]

    # Apply plateau models for specified population.
    ppo_expr = mu_expr * slope + intercept

    # Generate sum aggregators for 'predicted_proportion_observed' and
    # 'expected_variants', for specified population.
    agg_expr = {
        f"predicted_proportion_observed{pop}": agg_func(ppo_expr),
        f"expected_variants{pop}": agg_func(ppo_expr * cov_corr_expr),
    }

    # Generate sum aggregators for 'observed_variants' and 'possible_variants' on
    # the entire dataset if pop is None, and for `downsampling_counts` for
    # specified population if pop is not None.
    agg_expr.update({ann: agg_func(ht[ann]) for ann in ann_to_sum})

    return agg_expr

```

Reply:
I noticed that your code applied models on both observed and possible variants, while konrad's code only applied models to possible variants (see [here](https://github.com/broadinstitute/gnomad_lof/blob/e09d9fdc13d43c8ac996c93735ed206ba41ed473/constraint_utils/constraint_basics.py#L693)). Just for curiosity, was there a specific reason for this?